File: platforms/harmony/entry/src/main/ets/keyboard/input/QuickPunctuationPolicy.ts
/**
 * The quick punctuation menu for the alphabetic keyboard, ported from
 * platforms/android/java/app/msime/client/QuickPunctuationPolicy.java.
 *
 * Each entry prints one face and sends a different ASCII character: the Engine decides the shape from
 * the current scheme, so the host must send the key rather than the glyph.
 */
export interface PunctuationEntry {
  readonly face: string;
  /** The ASCII character handed to the Engine, not the printed face. */
  readonly input: string;
}

function entry(face: string, input: string): PunctuationEntry {
  const code: number = input.charCodeAt(0);
  if (face.length === 0 || code < 32 || code > 126) {
    throw new Error('Invalid quick punctuation entry');
  }
  return { face: face, input: input };
}

const ASCII: PunctuationEntry[] = [
  entry(',', ','), entry('.', '.'), entry('?', '?'), entry('!', '!'),
  entry(':', ':'), entry(';', ';'), entry('@', '@')
];
const CHINESE: PunctuationEntry[] = [
  entry('，', ','), entry('。', '.'), entry('？', '?'), entry('！', '!'),
  entry('、', '\\'), entry('；', ';'), entry('：', ':')
];
const JAPANESE: PunctuationEntry[] = [
  entry('、', '\\'), entry('。', '.'), entry('？', '?'), entry('！', '!'),
  entry('「', '['), entry('」', ']'), entry('・', '/')
];

export class QuickPunctuationPolicy {
  /** Display faces paired with the ASCII input each one sends to the Engine. */
  static entries(dedicatedEnglish: boolean, scheme: number, localMode: string): PunctuationEntry[] {
    if (dedicatedEnglish || localMode !== 'none') {
      return ASCII;
    }
    return scheme === 3 ? JAPANESE : CHINESE;
  }
}

File: platforms/android/java/app/msime/client/policy/QuickPunctuationPolicy.java
package app.msime.client;

import java.util.List;

/** Provides the Apple-compatible quick punctuation menu for the alphabetic keyboard. */
public final class QuickPunctuationPolicy {
    public record Entry(String face, char input) {
        public Entry {
            if (face == null || face.isEmpty() || input < 32 || input > 126)
                throw new IllegalArgumentException("Invalid quick punctuation entry");
        }
    }

    private static final List<Entry> ASCII = List.of(
        new Entry(",", ','), new Entry(".", '.'), new Entry("?", '?'), new Entry("!", '!'),
        new Entry(":", ':'), new Entry(";", ';'), new Entry("@", '@'));
    private static final List<Entry> CHINESE = List.of(
        new Entry("，", ','), new Entry("。", '.'), new Entry("？", '?'), new Entry("！", '!'),
        new Entry("、", '\\'), new Entry("；", ';'), new Entry("：", ':'));
    private static final List<Entry> JAPANESE = List.of(
        new Entry("、", '\\'), new Entry("。", '.'), new Entry("？", '?'), new Entry("！", '!'),
        new Entry("「", '['), new Entry("」", ']'), new Entry("・", '/'));

    private QuickPunctuationPolicy() {}

    /** Returns display faces and the ASCII input each entry sends to Engine. */
    public static List<Entry> entries(boolean dedicatedEnglish, int scheme, String localMode) {
        if (dedicatedEnglish || !"none".equals(localMode)) return ASCII;
        return scheme == 3 ? JAPANESE : CHINESE;
    }
}
